p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,483 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/connmgr"
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/pnet"
+	"github.com/libp2p/go-libp2p/core/sec"
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	"github.com/libp2p/go-libp2p/p2p/security/noise"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal/encoding"
+
+	logging "github.com/ipfs/go-log/v2"
+	ma "github.com/multiformats/go-multiaddr"
+	mafmt "github.com/multiformats/go-multiaddr-fmt"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multihash"
+	pionlogger "github.com/pion/logging"
+
+	"github.com/pion/webrtc/v3"
+)
+
+var log = logging.Logger("webrtc-transport")
+
+var dialMatcher = mafmt.And(mafmt.IP, mafmt.Base(ma.P_UDP), mafmt.Base(ma.P_WEBRTC), mafmt.Base(ma.P_CERTHASH))
+
+const (
+	// hansdhakeChannelNegotiated is used to specify that the
+	// handshake data channel does not need negotiation via DCEP.
+	// A constant is used since the `DataChannelInit` struct takes
+	// references instead of values.
+	hansdhakeChannelNegotiated = true
+	// handshakeChannelId is the agreed ID for the handshake data
+	// channel. A constant is used since the `DataChannelInit` struct takes
+	// references instead of values. We specify the type here as this
+	// value is only ever copied and passed by reference
+	handshakeChannelId = uint16(0)
+)
+
+// timeout values for the peerconnection
+// https://github.com/pion/webrtc/blob/v3.1.50/settingengine.go#L102-L109
+const (
+	DefaultDisconnectedTimeout = 5 * time.Second
+	DefaultFailedTimeout       = 25 * time.Second
+	DefaultKeepaliveTimeout    = 2 * time.Second
+)
+
+type WebRTCTransport struct {
+	webrtcConfig webrtc.Configuration
+	rcmgr        network.ResourceManager
+	privKey      ic.PrivKey
+	noiseTpt     *noise.Transport
+	localPeerId  peer.ID
+
+	// timeouts
+	peerConnectionTimeouts IceTimeouts
+
+	// in-flight connections
+	maxInFlightConnections uint32
+}
+
+var _ tpt.Transport = &WebRTCTransport{}
+
+type Option func(*WebRTCTransport) error
+
+type IceTimeouts struct {
+	Disconnect time.Duration
+	Failed     time.Duration
+	Keepalive  time.Duration
+}
+
+// WithPeerConnectionIceTimeouts sets the ice disconnect, failure and keepalive timeouts
+func WithPeerConnectionIceTimeouts(timeouts IceTimeouts) Option {
+	return func(t *WebRTCTransport) error {
+		if timeouts.Disconnect != 0 {
```

Comment:
Good question! Going to sync with @ckousik on this one!

Reply:
0 is not treated as a default, and instead disables the respective check (eg. if keepalive interval is 0, no keepalives are sent, or if disconnected timeout is 0, the connection never enters  a disconnected state). Refer [here](https://github.com/pion/ice/blob/v2.3.0/agent_config.go#L73-L85).

The timeouts are only set to their default values if they are not set in the settingEngine. Refer [here](https://github.com/pion/ice/blob/67f28cf23a8ae59a38e0128390ccdad27b5526be/agent_config.go#L213-L224)

Our implementation requires that the values be set (explicitly or implicitly) so that we can fail connections are free up resources if a remote disconnects abruptly, therefore 0 values are not acceptable. We also need to enforce that Failed timeout > disconnected timeout, and failed timeout > keepalive interval. (We currently enforce keepalive interval <= disconnected interval)